src/org/mozilla/javascript/NativeJSON.java

```diff
@@ -298,10 +298,10 @@ private static Object str(Object key, Scriptable holder,
             value = ((NativeJavaObject) value).unwrap();
             if (value instanceof Map) {
                 Map<?,?> map = (Map<?,?>) value;
-                NativeObject nObj = new NativeObject();
+                Scriptable nObj = state.cx.newObject(state.scope);
                 map.forEach((k, v) -> {
                     if (k instanceof CharSequence) {
-                        nObj.put(((CharSequence) k).toString(), nObj, v);
+                        nObj.put(((CharSequence) k).toString(), nObj, state.cx.getWrapFactory().wrap(state.cx, state.scope, v, v.getClass()));
```

Comment:
Should I be using `ScriptRuntime.toObject` instead of calling wrap directly? Or something else?

Reply:
It appears that Context.javaToJS does the right thing more than any of the alternatives I found. I didn't originally choose it because it's only available as a static method, and I already had a Context object, so I didn't think it needed to go through the process of finding the current Context.